be/src/exec/pipeline/scan/scan_operator.cpp

```diff
@@ -306,90 +293,57 @@ Status ScanOperator::_trigger_next_scan(RuntimeState* state, int chunk_source_in
     // to avoid holding mutex in bthread, we choose to initialize lazily here instead of in prepare
     if (is_uninitialized(_query_ctx)) {
         _query_ctx = state->exec_env()->query_context_mgr()->get(state->query_id());
     }
     starrocks::debug::QueryTraceContext query_trace_ctx = starrocks::debug::tls_trace_ctx;
     query_trace_ctx.id = reinterpret_cast<int64_t>(_chunk_sources[chunk_source_index].get());
     int32_t driver_id = CurrentThread::current().get_driver_id();
-    if (_workgroup != nullptr) {
-        workgroup::ScanTask task = workgroup::ScanTask(_workgroup, [wp = _query_ctx, this, state, chunk_source_index,
-                                                                    query_trace_ctx, driver_id](int worker_id) {
-            if (auto sp = wp.lock()) {
-                // Set driver_id here to share some driver-local contents.
-                // Current it's used by ExprContext's driver-local state
-                CurrentThread::current().set_pipeline_driver_id(driver_id);
-                DeferOp defer([]() { CurrentThread::current().set_pipeline_driver_id(0); });
-                SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(state->instance_mem_tracker());
-                [[maybe_unused]] std::string category = "chunk_source_" + std::to_string(chunk_source_index);
-                QUERY_TRACE_ASYNC_START("io_task", category, query_trace_ctx);
-                auto& chunk_source = _chunk_sources[chunk_source_index];
-                size_t num_read_chunks = 0;
-                int64_t prev_cpu_time = chunk_source->get_cpu_time_spent();
-                int64_t prev_scan_rows = chunk_source->get_scan_rows();
-                int64_t prev_scan_bytes = chunk_source->get_scan_bytes();
-
-                // Read chunk
-                Status status = chunk_source->buffer_next_batch_chunks_blocking_for_workgroup(
-                        kIOTaskBatchSize, state, &num_read_chunks, worker_id, _workgroup);
-                if (!status.ok() && !status.is_end_of_file()) {
-                    _set_scan_status(status);
-                }
-
-                int64_t delta_cpu_time = chunk_source->get_cpu_time_spent() - prev_cpu_time;
-                _workgroup->increment_real_runtime_ns(delta_cpu_time);
-                _workgroup->incr_period_scaned_chunk_num(num_read_chunks);
 
-                _finish_chunk_source_task(state, chunk_source_index, delta_cpu_time,
-                                          chunk_source->get_scan_rows() - prev_scan_rows,
-                                          chunk_source->get_scan_bytes() - prev_scan_bytes);
-                QUERY_TRACE_ASYNC_FINISH("io_task", category, query_trace_ctx);
+    workgroup::ScanTask task;
+    task.workgroup = _workgroup;
+    // TODO(by satanson): set a proper priority
+    task.priority = 20;
+    task.work_function = [wp = _query_ctx, this, state, chunk_source_index, query_trace_ctx, driver_id](int worker_id) {
+        if (auto sp = wp.lock()) {
```

Comment:
why here need a lock?

Reply:
This is the previous codes, and I just move the position of them.

`wp` is a `weak_ptr`, `wp.lock` is to check whether the `query_ctx` is still alive and get the `shared_ptr`, if it is alive.